pinc/Project.inc

```diff
@@ -955,83 +986,38 @@ class Project
 
     // -------------------------------------------------------------------------
 
-    // These should be treated as constants.
-    // (PHP doesn't appear to allow constants as class members.)
-    public $CBP_OKAY = 0;
-    public $CBP_PROJECT_NOT_IN_ROUND = 1;
-    public $CBP_PROJECT_NOT_AVAILABLE = 2;
-    public $CBP_REQUESTOR_NOT_LOGGED_IN = 3;
-    public $CBP_USER_NOT_QUALIFIED_FOR_ROUND = 4;
-    public $CBP_PROJECT_REQUIRES_MAINTENANCE = 5;
-
-    public function can_be_proofed_by_current_user()
+    public function validate_can_be_proofed_by_current_user()
     // (where "proofed" means "worked on in a round, right now".)
-    // Returns an array consisting of:
-    // -- one of the above codes, and
-    // -- a suggested error message.
+    // Can throw Exceptions
     {
-        global $code_url, $pguser;
+        global $pguser;
 
-        $state = $this->state;
-        $round = get_Round_for_project_state($state);
+        $project_round = $this->get_project_available_round();
 
-        if (is_null($round)) {
-            // The project is not in any round.
-            return [
-                $this->CBP_PROJECT_NOT_IN_ROUND,
-                _('The project is not in a round.'),
-            ];
+        if (is_null($pguser)) {
+            throw new UserNotLoggedInException(_('Registered users can work on the project, but you are not logged in.'));
         }
 
+        $project_round->validate_user_can_access($pguser);
+    }
+
+    public function get_project_available_round()
+    {
+        $state = $this->state;
+        $project_round = get_Round_for_project_state($state);
+        if (is_null($project_round)) {
+            // The project is not in any round.
+            throw new ProjectNotInRoundException(_("The project is not in a round"));
+        }
         if (!$this->is_utf8) {
             // Project has not been converted to UTF-8 yet
-            return [
-                $this->CBP_PROJECT_REQUIRES_MAINTENANCE,
-                _('The project requires maintenance.') . ' ' . _("Project table is not UTF-8."),
-            ];
+            throw new ProjectRequiresMaintenanceException(_('The project requires maintenance. Project table is not UTF-8.'));
```

Comment:
> These can be fixed after the testing is done, but for now, if anyone else wants to verify, `projectID459698b8b63ac` has been converted to `latin1` (so has `projectID6229978d368ca`, but it's a new project, and the blurb box with the "requires maintenance" message doesn't appear for new projects; just the error message in the project data on the project page).

It has moved into `Proofreading Round 1: Available` now, and the blurb box appears with the message. But the link to start proofreading does not appear. It looks the same on DPT branch so it looks like nothing is broken.

Reply:
Sorry for the confusion. There were two projects that I converted to `latin1`. The first one was already in P1 Available, and showed the blurb box. The second one was/is still a new project, and doesn't show the blurb box because the blurb box isn't shown in that case.

Reply:
Probably my confusion.